### PR TITLE
main (test): integrate test corpus runner

### PR DESCRIPTION
**Pull request by niaow**
This allows us to test a large corpus of external packages against the compiler.

This is based off of https://github.com/dgryski/tinygo-test-corpus

There are still a few things I need to split out before this is ready

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -262,6 +262,14 @@ tinygo-test-wasi:
 tinygo-bench-wasi:
 	$(TINYGO) test -target wasi -bench . $(TEST_PACKAGES_WASI)
 
+# Test external packages in a large corpus.
+test-corpus: wasi-libc
```
Comment (dgryski) [inline Makefile]: regular corpus probably doesn't need `wasi-libc` as a dependency.

**Comment by dkegel-fastly** [comment]
Should the corpus yaml go in testdata?

**Comment by niaow** [comment]
Currently waiting on #2520 and #2517

**Comment by dgryski** [comment]
Now that https://github.com/tinygo-org/tinygo/pull/2455 is in you can update the list of skipwasi packages:

https://github.com/dgryski/tinygo-test-corpus/commit/759365bd1614cd583424b2102db770be36d1017b

File: corpus.yaml [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+  subdirs:
+  # - bmp: needs  _time.startTimer
+  - pkg: ccitt
+  - pkg: colornames
+  - pkg: draw
+  - pkg: font
+  - pkg: font/basicfont
+  - pkg: font/opentype
+  - pkg: font/plan9font
+    skipwasi: true # 2022/01/10 10:43:31 === RUN   TestMetrics: panic: (reflect.Value).Interface: unexported
```
Comment (dgryski) [inline corpus.yaml]: This was also file access.

**Comment by niaow** [comment]
This has been rebased now that #2520 and #2517 have been merged. I also added a `test-corpus-fast` which skips the few extremely slow tests.

File: testdata/corpus.yaml [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+  subdirs:
+  # - bmp: needs  _time.startTimer
+  - pkg: ccitt
+  - pkg: colornames
+  - pkg: draw
+  - pkg: font
+  - pkg: font/basicfont
+  - pkg: font/opentype
+  - pkg: font/plan9font
+    skipwasi: true # 2022/01/10 10:43:31 === RUN   TestMetrics: panic: (reflect.Value).Interface: unexported
```
Comment (dgryski) [inline testdata/corpus.yaml]: Turns out this was also file access.  This test should pass on wasi.

File: testdata/corpus.yaml [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+  - pkg: integrate/quad
+  - pkg: internal/cmplx64
+  - pkg: internal/testrand
+  - pkg: lapack/gonum
+    skipwasi: true # takes too long
+    slow: true
+  - pkg: mathext
+  - pkg: mathext/prng
+  - pkg: optimize/convex/lp
+    skipwasi: true # takes too long
```
Comment (dgryski) [inline testdata/corpus.yaml:225]: Need a `slow: true` here too or is it fast (enough) on native?

File: testdata/corpus.yaml [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+  - pkg: spatial/r3
+  - pkg: stat/distmat
+  - pkg: stat/mds
+  - pkg: stat/spatial
+  - pkg: stat/distmv
+    skipwasi: true # takes too long
+    slow: true
+  - pkg: stat/samplemv
+    skipwasi: true # takes too long
+            # 'blas/blas64', # -- TestDasum panic: blas: n < 0
```
Comment (dgryski) [inline testdata/corpus.yaml:237]: `slow: true` or fast enough on native?

File: corpus_test.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+			out := ioLogger(t, &wg)
+			defer out.Close()
+
+			dir := t.TempDir()
+			cmd := exec.Command("go", "mod", "init", "github.com/tinygo/tinygo-corpus-test")
+			cmd.Dir = dir
+			cmd.Stdout, cmd.Stderr = out, out
+			err := cmd.Run()
+			if err != nil {
+				t.Errorf("failed to init: %s", err.Error())
```
Comment (dgryski) [inline corpus_test.go:89]: `%s` should call `Error()` for error types already.

**Review by dgryski — CHANGES_REQUESTED** [review]

File: testdata/corpus.yaml (hunk 4 above)
Comment (niaow) [reply testdata/corpus.yaml:225]: Fast enough on native

File: testdata/corpus.yaml (hunk 5 above)
Comment (niaow) [reply testdata/corpus.yaml:237]: Fast enough on native

File: corpus_test.go (hunk 6 above)
Comment (niaow) [reply corpus_test.go:89]: Yes but I prefer doing it explicitly so that I know what it is doing.

**Review by dgryski — APPROVED** [review]
LGTM

File: Makefile [hunk 7]
```diff
@@ -295,6 +295,15 @@ tinygo-bench-wasi:
 tinygo-bench-wasi-fast:
 	$(TINYGO) test -target wasi -bench . $(TEST_PACKAGES_WASI)
 
+# Test external packages in a large corpus.
+test-corpus:
+	CGO_CPPFLAGS="$(CGO_CPPFLAGS)" CGO_CXXFLAGS="$(CGO_CXXFLAGS)" CGO_LDFLAGS="$(CGO_LDFLAGS)" $(GO) test $(GOTESTFLAGS) -timeout=1h -buildmode exe -tags byollvm -run TestCorpus . -corpus=testdata/corpus.yaml
+test-corpus-fast:
+	CGO_CPPFLAGS="$(CGO_CPPFLAGS)" CGO_CXXFLAGS="$(CGO_CXXFLAGS)" CGO_LDFLAGS="$(CGO_LDFLAGS)" $(GO) test $(GOTESTFLAGS) -timeout=1h -buildmode exe -tags byollvm -run TestCorpus -short . -corpus=testdata/corpus.yaml
+test-corpus-wasi: wasi-libc
+	CGO_CPPFLAGS="$(CGO_CPPFLAGS)" CGO_CXXFLAGS="$(CGO_CXXFLAGS)" CGO_LDFLAGS="$(CGO_LDFLAGS)" $(GO) test $(GOTESTFLAGS) -timeout=1h -buildmode exe -tags byollvm -run TestCorpus . -corpus=testdata/corpus.yaml -target=wasi
```
Comment (dkegel-fastly) [inline Makefile:304]: may want test-corpus-wasi-fast as well.

Comment (niaow) [reply Makefile:304]: I guess I could, but all tests tagged as slow are currently skipwasi

Comment (dkegel-fastly) [reply Makefile:304]: Oh, never mind, then!

**Comment by deadprogram** [comment]
Please resolve merge conflict @niaow thanks!

**Comment by niaow** [comment]
Done!

**Pull request merged**